pkg/user/user.go

```diff
@@ -0,0 +1,190 @@
+		parseString(line, &up.user, &up.placeholder, &up.uid)
+		if (idErr == nil && idInt == up.uid) || str == up.user {
+			return uint32(up.uid), true
+		}
+		return 0, false
+	})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// if gidStr is null, then get gid from /etc/passwd
+	if len(gidStr) == 0 {
+		gid, err = parseID(filepath.Join(path, passwdFile), uidStr, func(line, str string, idInt int, idErr error) (uint32, bool) {
+			var up uidParser
+			parseString(line, &up.user, &up.placeholder, &up.uid, &up.gid)
+			if (idErr == nil && idInt == up.uid) || str == up.user {
+				return uint32(up.gid), true
+			}
+			return 0, false
+		})
+		if err != nil {
+			return 0, 0, err
+		}
+	} else {
+		gid, err = parseID(filepath.Join(path, groupFile), gidStr, func(line, str string, idInt int, idErr error) (uint32, bool) {
+			var gp gidParser
+			parseString(line, &gp.group, &gp.placeholder, &gp.gid)
+			if (idErr == nil && idInt == gp.gid) || str == gp.group {
+				return uint32(gp.gid), true
+			}
+			return 0, false
+		})
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+
+	return uid, gid, nil
+}
+
+// GetIntegerID only parser user format uid:gid, cause container rootfs is not created
+// by contianerd now, can not change user to id, only support user id >= 1000
+func GetIntegerID(user string) (uint32, uint32) {
+	if user == "" {
+		// return default user root
+		return 0, 0
+	}
+
+	// if uid gid can not be parsed successfully, return default user root
+	var uid, gid int
+	parseString(user, &uid, &gid)
+	return uint32(uid), uint32(gid)
+}
+
+// parseID parses uid from /etc/passwd.
+func parseID(file, str string, parserFilter filterFunc) (uint32, error) {
+	idInt, idErr := strconv.Atoi(str)
+
+	ba, err := ioutil.ReadFile(file)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read passwd file %s: %s", passwdFile, err)
+	}
+
+	scanner := bufio.NewScanner(bytes.NewReader(ba))
+	for scanner.Scan() {
+		line := scanner.Text()
+		id, ok := parserFilter(line, str, idInt, idErr)
+		if ok {
+			return id, nil
+		}
+	}
+
+	if idErr == nil {
+		ok, err := isUnknowUser(idInt)
+		if err != nil {
+			return 0, err
+		}
+		if ok {
+			return uint32(idInt), nil
+		}
+	}
+
+	return 0, fmt.Errorf("failed to find id or name %s in %s", str, file)
+}
+
+// isUnknowUser determines if id can be accepted as a unknow user. this kind of user id should >= 1000
+func isUnknowUser(id int) (bool, error) {
+	// first test id is valid in minID ~ maxID
+	if id < minID || id > maxID {
+		return false, fmt.Errorf("use id %d out of range, should be in %d ~ %d", id, minID, maxID)
+	}
+
+	if id < acceptedID {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// parseString parses line in format a:b:c.
+func parseString(line string, v ...interface{}) {
+	splits := strings.Split(line, ":")
+	for i, s := range splits {
+		if len(v) <= i {
+			break
+		}
+
+		switch p := v[i].(type) {
+		case *string:
+			*p = s
+		case *int:
+			*p, _ = strconv.Atoi(s)
+		case *[]string:
```

Comment:
I suggest that we could add test case in the unit test for this case.

Reply:
already get `TestParseString` ut in test file